src/Datadog.Trace.DuckTyping/DuckType.Methods.cs

```diff
@@ -0,0 +1,605 @@
+                    // we can't access non public types so we have to cast to object type (in the instance object and the return type).
+
+                    string dynMethodName = $"_callMethod+{targetMethod.DeclaringType.Name}.{targetMethod.Name}";
+                    returnType = (targetMethod.ReturnType.IsPublic || targetMethod.ReturnType.IsNestedPublic) && !targetMethod.ReturnType.IsGenericParameter ? targetMethod.ReturnType : typeof(object);
+
+                    // We create the dynamic method
+                    Type[] originalTargetParameters = targetMethod.GetParameters().Select(p => p.ParameterType).ToArray();
+                    Type[] targetParameters = targetMethod.IsStatic ? originalTargetParameters : (new[] { typeof(object) }).Concat(originalTargetParameters).ToArray();
+                    Type[] dynParameters = targetMethod.IsStatic ? targetMethodParametersTypes : (new[] { typeof(object) }).Concat(targetMethodParametersTypes).ToArray();
+                    DynamicMethod dynMethod = new DynamicMethod(dynMethodName, returnType, dynParameters, typeof(DuckType).Module, true);
+
+                    // We store the dynamic method in a bag to avoid getting collected by the GC.
+                    DynamicMethods.Add(dynMethod);
+
+                    // Emit the dynamic method body
+                    ILGenerator dynIL = dynMethod.GetILGenerator();
+
+                    if (!targetMethod.IsStatic)
+                    {
+                        ILHelpers.LoadInstanceArgument(dynIL, typeof(object), targetMethod.DeclaringType);
+                    }
+
+                    for (int idx = targetMethod.IsStatic ? 0 : 1; idx < dynParameters.Length; idx++)
+                    {
+                        ILHelpers.WriteLoadArgument(idx, dynIL, true);
+                        ILHelpers.TypeConversion(dynIL, dynParameters[idx], targetParameters[idx]);
+                    }
+
+                    // Check if we can emit a normal Call/CallVirt to the target method
+                    if (!targetMethod.ContainsGenericParameters)
+                    {
+                        dynIL.EmitCall(targetMethod.IsStatic ? OpCodes.Call : OpCodes.Callvirt, targetMethod, null);
+                    }
+                    else
+                    {
+                        // We can't emit a call to a method with generics from a DynamicMethod
+                        // Instead we emit a Calli with the function pointer.
+                        dynIL.Emit(OpCodes.Ldc_I8, (long)targetMethod.MethodHandle.GetFunctionPointer());
+                        dynIL.Emit(OpCodes.Conv_I);
+                        dynIL.EmitCalli(
+                            OpCodes.Calli,
+                            targetMethod.CallingConvention,
+                            targetMethod.ReturnType,
+                            targetMethod.GetParameters().Select(p => p.ParameterType).ToArray(),
+                            null);
```

Comment:
This seems to me like a good candidate for wrapping into an `ILHelpers` method with the `MethodInfo` object. Same with an overload for `DynamicMethod` below.

Reply:
Yeah, I like the idea, I'll refactor all the `Calli`'s

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/946/commits/e3020964dcbb2b80b94368777abbdd5b1de4e248